message/router/middleware/deduplicator.go

```diff
@@ -0,0 +1,237 @@
+			delete(d.tags, hash)
+		}
+	}
+}
+
+// Len returns the number of known tags that have not been
+// cleaned out yet.
+func (d *Deduplicator) Len() (count int) {
+	d.mu.Lock()
+	count = len(d.tags)
+	d.mu.Unlock()
+	return
+}
+
+// IsDuplicate returns true if the message hash tag calculated
+// using a [MessageHasher] was seen in deduplication time window.
+func (d *Deduplicator) IsDuplicate(m *message.Message) (bool, error) {
+	tag, err := d.tagger(m)
+	if err != nil {
+		return false, err
+	}
+
+	d.mu.Lock()
+	_, alreadySeen := d.tags[tag]
+	if alreadySeen {
+		// NOTE: could also check if tag expires.After(t)
+		// and remove it for exact expiration
+		// instead of fuzzy until-next clean up expiration
+		// but this should not be needed for most use cases.
+		d.mu.Unlock()
+		return true, nil
+	}
+	d.tags[tag] = time.Now().Add(d.window)
+	d.mu.Unlock()
+	return false, nil // first time, not a duplicate
+}
+
+// NewDeduplicator returns a new Deduplicator.
+// Call [Deduplicator.Middleware] for a new middleware
+// or [Deduplicator.Decorator] for a [message.PublisherDecorator].
+//
+// Use [NewMessageHasherAdler32] for fast tagging or
+// [NewMessageHasherSHA256] for minimal collisions.
+//
+// Window specifies the minimum duration of  how long the
+// duplicate tags are remembered for. Real duration can
+// extend up to 50% longer because it depends on the
+// clean up cycle.
+func NewDeduplicator(
+	tagger MessageHasher,
+	window time.Duration,
+) *Deduplicator {
+	if tagger == nil {
+		panic("cannot use a <nil> tagger")
+	}
+	if window < time.Millisecond {
+		panic("deduplication window of less than a millisecond is impractical")
+	}
+
+	d := &Deduplicator{
+		tagger: tagger,
+		window: window,
+
+		mu:   &sync.Mutex{},
+		tags: make(map[string]time.Time),
+	}
+	go d.cleanOutLoop(context.Background(), time.NewTicker(window/2))
+	return d
+}
+
+// Middleware returns the [Deduplicator] middleware.
+func (d *Deduplicator) Middleware(h message.HandlerFunc) message.HandlerFunc {
+	return func(msg *message.Message) ([]*message.Message, error) {
+		isDuplicate, err := d.IsDuplicate(msg)
+		if err != nil {
+			return nil, err
+		}
+		if isDuplicate {
+			return nil, nil
+		}
+		return h(msg)
+	}
+}
+
+type deduplicatingPublisherDecorator struct {
+	message.Publisher
+	deduplicator *Deduplicator
+}
+
+func (d *deduplicatingPublisherDecorator) Publish(
+	topic string,
+	messages ...*message.Message,
+) (err error) {
+	notRecent := make([]*message.Message, 0, len(messages))
+	isDuplicate := false
+
+	for _, m := range messages {
+		isDuplicate, err = d.deduplicator.IsDuplicate(m)
+		if err != nil {
+			return err
+		}
+		if isDuplicate {
+			m.Ack() // acknowledge and ignore
+			continue
+		}
+		notRecent = append(notRecent, m)
+	}
+	return d.Publisher.Publish(topic, notRecent...)
+}
+
+// PublisherDecorator returns a decorator that
+// acknowledges and drops every [message.Message] that
+// was recognized by a [Deduplicator].
```

Comment:
it would be good to explain use case for that

I would also put it a bit up, as it's entrypoint of the public API

Reply:
Adding:

    // The returned decorator provides the same functionality
    // to a [message.Publisher] as [Deduplicator.Middleware]
    // to a [message.Router].